File: camera/Attic/blob/Blob.cpp
#include "Blob.h"

Blob::Blob(int w, int h, int d) {
  initialize(w, h, 3, 0, 1, 2);
}

void Blob::drawRect(long int left, long int right, 
		    long int top, long int bottom, 
		    long int r, long int g, long int b) {
  for(int w=left; w<=right; w++) {
    for(int h=top; h<=bottom; h++ ) {
      image[(h * width + w) * depth + rgb[0]] = r;
      image[(h * width + w) * depth + rgb[1]] = g;
      image[(h * width + w) * depth + rgb[2]] = b;
    }
  }
}

void Blob::clear() {
  for(int w=0; w<width; w++) {
    for(int h=0; h<height; h++ ) {
      image[(h * width + w) * depth + rgb[0]] = 128;
      image[(h * width + w) * depth + rgb[1]] = 128;
      image[(h * width + w) * depth + rgb[2]] = 128;
    }
  }
}

PyObject *Blob::updateMMap(PyObject *blobData) {
  int r, g, b;
  PyObject *blob;
  clear();
  if (!PyTuple_Check(blobData)) {
    PyErr_SetString(PyExc_TypeError, "Invalid tuple to updateMMap()");
    return NULL;
  }
  long int packedcolor, area, center_x, center_y, 
    left, right, top, bottom, dist;
  for (int i = 0; i < PyTuple_Size(blobData); i++) {
    PyObject *blobTuple = PyTuple_GetItem(blobData, i);
    // for each blob:
    for (int blobcount = 0; blobcount < PyTuple_Size(blobTuple); blobcount++) {
      if (!PyArg_ParseTuple(blobTuple, "O", &blob)) {
	PyErr_SetString(PyExc_TypeError, "Invalid tuple to updateMMap()");
	return NULL;
      }
      if (!PyArg_ParseTuple(blob, "lllllllll", 
			    &packedcolor, &area, &center_x, &center_y, \
			    &left, &right, &top, &bottom, &dist)) {
	PyErr_SetString(PyExc_TypeError, "Invalid items to updateMMap()");
	return NULL;
      }
      // process data
      // draw rectangle
      r = packedcolor >> 16;
      g = packedcolor >> 8 & 0x0000FF;
      b = packedcolor & 0x0000FF;
      drawRect(left, right, top, bottom, r, g, b);
    }
  }
  return PyInt_FromLong(0L);
}

File: camera/blob.py
from pyro.camera import *
from struct import pack
import ImageDraw

class BlobCamera(Camera):
   """
   """
   def __init__(self, robot):
      if not robot.hasService('blobfinder'):
         raise "Robot does not have service", 'blobfinder'
      self.robot = robot
      self.blobdata = []
      # do this last:
      Camera.__init__(self, 0, 0)
      self.colorpos = { (255, 255, 255) : 0,
                        (255, 0, 0): 1,
                        (0, 255, 0) : 2, 
                        (0, 0, 255) : 3,
                        (0, 255, 255) : 4,
                        (255, 255, 0) : 5, 
                        (255, 0, 255) : 6, 
                        (0, 0, 0) : 7}
      
   def _update(self):
      self.blobdata = self.robot.getServiceData('blobfinder')

   def getImage(self):
      self.width, self.height = self.blobdata[0]
      # mode, (x, y), color (r, g, b): 0 - 255
      image = PIL.PpmImagePlugin.Image.new('P', (self.width, self.height), 0)
      image.putpalette([
         255, 255, 255, # background
         255, 0, 0,   # red
         0, 255, 0,   # green
         0, 0, 255,   # blue
         0, 255, 255, # yellow
         255, 255, 0,   # red
         255, 0, 255,   # red
         0, 0, 0,   # red
         ])
      draw = ImageDraw.ImageDraw(image)
      for i in range(32): # each color channel
         for blobcount in range(len(self.blobdata[1][i])): # for each blob
            packedcolor, area, center_x, center_y, \
                         left, right, top, bottom, \
                         dist = self.blobdata[1][i][blobcount]
            r = packedcolor >> 16
            g = packedcolor >> 8 & 0x0000FF
            b = packedcolor & 0x0000FF
            color = self.colorpos[(r, g, b)]
            draw.setfill(color)
            draw.setink(color)
            # draw in the image area
            draw.rectangle((left, top, right, bottom))
      return image
